File: app/src/main/java/co/medellinandroid/spotifyartists/topTenArtists/TopTenArtistsPresenter.java
package co.medellinandroid.spotifyartists.topTenArtists;

import co.medellinandroid.spotifyartists.model.SpotifyArtistsResponse;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import javax.inject.Inject;

public class TopTenArtistsPresenter implements TopTenArtistsContract.Presenter {

  private TopTenArtistsContract.View view;
  private TopTenArtistsInteractor topTenArtistsInteractor;
  private DisposableObserver<SpotifyArtistsResponse> disposableObserver;

  @Inject public TopTenArtistsPresenter() {
    this.topTenArtistsInteractor = new TopTenArtistsInteractor();
  }

  @Override public void attachView(TopTenArtistsContract.View view) {
    this.view = view;
  }

  @Override public void getTopTenArtists() {
    if (null != view) view.showProgress();
    disposableObserver = new DisposableObserver<SpotifyArtistsResponse>() {
      @Override public void onNext(SpotifyArtistsResponse spotifyArtistsResponse) {
        view.hideProgress();
        view.showList(spotifyArtistsResponse.getArtists().getItems());
      }

      @Override public void onError(Throwable e) {
        view.hideProgress();
        view.showError(e.getMessage());
      }

      @Override public void onComplete() {
        view.hideProgress();
      }
    };

    topTenArtistsInteractor.getTopTenArtists()
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(disposableObserver);
  }

  @Override public void onDestroy() {
    view = null;
    if (null != disposableObserver && !disposableObserver.isDisposed()) {
      disposableObserver.dispose();
    }
  }
}

File: app/src/main/java/co/medellinandroid/spotifyartists/topTenArtists/TopTenArtistsPresenter.kt
package co.medellinandroid.spotifyartists.topTenArtists

import co.medellinandroid.spotifyartists.model.SpotifyArtistsResponse
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

class TopTenArtistsPresenter @Inject constructor() : TopTenArtistsContract.Presenter {

  internal var view: TopTenArtistsContract.View? = null
  private val topTenArtistsInteractor: TopTenArtistsInteractor = TopTenArtistsInteractor()
  private lateinit var disposableObserver: DisposableObserver<SpotifyArtistsResponse>

  override fun attachView(view: TopTenArtistsContract.View) {
    this.view = view
  }

  override fun getTopTenArtists() {
    view?.showProgress()
    disposableObserver = object : DisposableObserver<SpotifyArtistsResponse>() {
      override fun onNext(spotifyArtistsResponse: SpotifyArtistsResponse) {
        view?.hideProgress()
        view?.showList(spotifyArtistsResponse.artists!!.items)
      }

      override fun onError(e: Throwable) {
        view?.hideProgress()
        view?.showError(e.message!!)
      }

      override fun onComplete() {
        view?.hideProgress()
      }
    }

    topTenArtistsInteractor.topTenArtists
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(disposableObserver)
  }

  override fun onDestroy() {
    view = null
    if (!disposableObserver.isDisposed) {
      disposableObserver.dispose()
    }
  }
}
